File: Converter/Scripts/GhidraDecompile.py
# GhidraDecompile.py
# ─────────────────────────────────────────────────────────────────────────────
# Ghidra headless post-script.
# Reads the manifest CSV written by Il2CppExplorer and decompiles each
# function to a C source file.
#
# Manifest format (CSV, first line = header):
#   Address,OutputFile,Length,TypeName,MemberName
#   0x210B70,"C:\output\Ns\Class\ctor.c",0x7,SomeNamespace.SomeClass,.ctor
#
# The Address column contains either the RVA or file offset (controlled by
# the --use-offset flag in Il2CppExplorer).
#
# Ghidra uses this script with:
#   analyzeHeadless ... -postScript GhidraDecompile "<manifest_path>"
# ─────────────────────────────────────────────────────────────────────────────

import os
import sys

# Ghidra API imports (available inside the Ghidra scripting environment)
from ghidra.app.decompiler import DecompInterface, DecompileOptions
from ghidra.util.task import ConsoleTaskMonitor
from ghidra.program.model.address import AddressOutOfBoundsException

# ── Helpers ───────────────────────────────────────────────────────────────────

def parse_hex(s):
    """Convert '0x1A2B' or '1A2B' to an integer; return 0 on failure."""
    s = s.strip().strip('"')
    if not s or s == '0':
        return 0
    try:
        return int(s, 16)
    except ValueError:
        return 0

def parse_manifest(path):
    """
    Parse the CSV manifest.  Returns a list of dicts with keys:
      address_str, output_file, length_str, type_name, member_name
    Skips the header row and blank lines.
    """
    rows = []
    with open(path, 'r') as f:
        lines = f.readlines()
    for line in lines[1:]:          # skip header
        line = line.rstrip('\r\n')
        if not line:
            continue
        # Handle quoted OutputFile which may contain commas
        if ',"' in line:
            first_comma = line.index(',')
            rest = line[first_comma + 1:]   # starts with "
            closing_quote = rest.index('"', 1)
            out_file = rest[1:closing_quote]
            tail = rest[closing_quote + 2:] # skip closing quote + comma
            parts = tail.split(',', 2)
            rows.append({
                'address_str': line[:first_comma],
                'output_file': out_file,
                'length_str':  parts[0] if len(parts) > 0 else '0',
                'type_name':   parts[1] if len(parts) > 1 else '',
                'member_name': parts[2] if len(parts) > 2 else '',
            })
        else:
            parts = line.split(',', 4)
            if len(parts) < 3:
                continue
            rows.append({
                'address_str': parts[0],
                'output_file': parts[1],
                'length_str':  parts[2],
                'type_name':   parts[3] if len(parts) > 3 else '',
                'member_name': parts[4] if len(parts) > 4 else '',
            })
    return rows

def ensure_dir(path):
    d = os.path.dirname(path)
    if d and not os.path.exists(d):
        os.makedirs(d)

def write_file(path, content):
    ensure_dir(path)
    with open(path, 'w') as f:
        f.write(content)

# ── Main script entry point ───────────────────────────────────────────────────

def run():
    args = getScriptArgs()
    if not args:
        print("ERROR: GhidraDecompile.py requires the manifest CSV path as a script argument.")
        print("  Usage: -postScript GhidraDecompile \"<manifest_path>\"")
        return

    manifest_path = args[0].strip('"').strip("'")
    if not os.path.exists(manifest_path):
        print("ERROR: Manifest not found: " + manifest_path)
        return

    print("GhidraDecompile: reading manifest from " + manifest_path)
    rows = parse_manifest(manifest_path)
    print("GhidraDecompile: {} entries to decompile".format(len(rows)))

    # Set up decompiler
    options = DecompileOptions()
    decompiler = DecompInterface()
    decompiler.setOptions(options)
    decompiler.setSimplificationStyle("decompile")
    decompiler.openProgram(currentProgram)

    monitor = ConsoleTaskMonitor()

    image_base = currentProgram.getImageBase().getOffset()
    print("GhidraDecompile: image base = 0x{:X}".format(image_base))

    ok_count   = 0
    fail_count = 0

    for i, row in enumerate(rows):
        addr_int    = parse_hex(row['address_str'])
        out_file    = row['output_file']
        type_name   = row['type_name']
        member_name = row['member_name']

        if addr_int == 0:
            write_file(out_file,
                "// Skipped: address is 0\n"
                "// Type  : {}\n// Member: {}\n".format(type_name, member_name))
            fail_count += 1
            continue

        # Resolve address
        # The address in the manifest may be:
        #   (a) RVA — relative to image base:  absolute = imageBase + rva
        #   (b) File offset — passed as-is to getFunctionAt (works if Ghidra
        #       is configured to load at file-offset addresses; less common).
        # We attempt imageBase+addr first; if not mapped we fall back to addr.

        flat_addr = image_base + addr_int
        try:
            ghidra_addr = toAddr(flat_addr)
        except Exception:
            ghidra_addr = None

        if ghidra_addr is None or not currentProgram.getMemory().contains(ghidra_addr):
            # Try raw addr (file offset or already-absolute address)
            try:
                ghidra_addr = toAddr(addr_int)
            except Exception:
                ghidra_addr = None

        if ghidra_addr is None:
            msg = ("// ERROR: could not resolve address {}\n"
                   "// Type  : {}\n// Member: {}\n").format(row['address_str'], type_name, member_name)
            write_file(out_file, msg)
            fail_count += 1
            print("  [{}/{}] ADDR_ERR {} :: {}".format(i + 1, len(rows), type_name, member_name))
            continue

        # Get or create the function
        func = getFunctionAt(ghidra_addr)
        if func is None:
            func = createFunction(ghidra_addr, None)

        if func is None:
            msg = ("// ERROR: could not find/create function at {}\n"
                   "// Type  : {}\n// Member: {}\n").format(row['address_str'], type_name, member_name)
            write_file(out_file, msg)
            fail_count += 1
            print("  [{}/{}] NO_FUNC {} :: {}".format(i + 1, len(rows), type_name, member_name))
            continue

        # Decompile
        result = decompiler.decompileFunction(func, 60, monitor)

        if result.decompileCompleted():
            c_code = result.getDecompiledFunction().getC()
            header = (
                "// Type  : {}\n"
                "// Member: {}\n"
                "// RVA   : {}\n"
                "// ────────────────────────────────────────\n\n"
            ).format(type_name, member_name, row['address_str'])
            write_file(out_file, header + c_code)
            ok_count += 1
            print("  [{}/{}] OK  {} :: {}".format(i + 1, len(rows), type_name, member_name))
        else:
            err_msg = result.getErrorMessage() or "unknown error"
            msg = (
                "// Decompilation failed\n"
                "// Type  : {}\n"
                "// Member: {}\n"
                "// RVA   : {}\n"
                "// Error : {}\n"
            ).format(type_name, member_name, row['address_str'], err_msg)
            write_file(out_file, msg)
            fail_count += 1
            print("  [{}/{}] FAIL {} :: {} — {}".format(i + 1, len(rows), type_name, member_name, err_msg))

    decompiler.dispose()
    print("\nGhidraDecompile: done. OK={} FAIL={}".format(ok_count, fail_count))

run()

File: Converter/Scripts/GhidraDecompile.java
// GhidraDecompile.java
// Ghidra headless post-script — reads the Il2CppExplorer manifest CSV and
// decompiles each listed function to a C source file.
//
// Manifest format (CSV, first line = header):
//   Address,OutputFile,Length,TypeName,MemberName
//   0x210B70,"C:\output\Ns\Class\ctor.c",0x7,SomeNamespace.SomeClass,.ctor
//
// Usage:
//   analyzeHeadless ... -postScript GhidraDecompile "<manifest_path>"
//
//@author Il2CppExplorer
//@category Analysis
//@keybinding
//@menupath
//@toolbar

import java.io.*;
import java.nio.file.*;
import java.util.*;

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileOptions;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.util.task.ConsoleTaskMonitor;
import ghidra.program.model.symbol.SourceType;

public class GhidraDecompile extends GhidraScript {

    @Override
    public void run() throws Exception {
        String[] scriptArgs = getScriptArgs();
        if (scriptArgs == null || scriptArgs.length == 0) {
            printerr("GhidraDecompile: manifest path argument is required.");
            printerr("  Usage: -postScript GhidraDecompile \"<manifest_path>\"");
            return;
        }

        String manifestPath = scriptArgs[0].replace("\\\\", "\\");
        File manifestFile = new File(manifestPath);
        if (!manifestFile.exists()) {
            printerr("GhidraDecompile: manifest not found: " + manifestPath);
            return;
        }

        println("GhidraDecompile: reading manifest from " + manifestPath);
        List<String[]> rows = parseManifest(manifestFile);
        println("GhidraDecompile: " + rows.size() + " entries to decompile");

        long imageBase = currentProgram.getImageBase().getOffset();
        println("GhidraDecompile: image base = 0x" + Long.toHexString(imageBase).toUpperCase());

        // ── Optional rename pass: label all known FUN_ addresses with C# names ──
        // This makes cross-calls readable: FUN_1808e20f0 → BattleTeam__ctor
        if (scriptArgs.length > 1) {
            String labelsPath = scriptArgs[1].replace("\\\\", "\\");
            File labelsFile = new File(labelsPath);
            if (labelsFile.exists()) {
                println("GhidraDecompile: renaming functions from labels file...");
                List<String[]> labels = parseLabels(labelsFile);
                int renamed = 0;
                for (String[] row : labels) {
                    long rva = parseHex(row[0]);
                    if (rva == 0) continue;
                    Address addr = resolveAddress(rva, imageBase);
                    if (addr == null) continue;
                    Function func = getFunctionAt(addr);
                    if (func == null) continue; // only rename what Ghidra already found
                    try {
                        func.setName(row[1].trim(), SourceType.USER_DEFINED);
                        renamed++;
                    } catch (Exception ignored) {}
                }
                println("GhidraDecompile: renamed " + renamed + " / " + labels.size() + " functions");
            }
        }

        // ── Optional static data labels pass: name class statics pointers ──────
        // scriptArgs[2] = path to _static_labels.csv  (RVA,Label — same format as _labels.csv)
        // Applies named symbols to DAT_ data addresses so the decompiler emits
        // readable names instead of DAT_XXXXXXXX in the next decompile run.
        if (scriptArgs.length > 2) {
            String staticLabelsPath = scriptArgs[2].replace("\\\\", "\\");
            File staticLabelsFile = new File(staticLabelsPath);
            if (staticLabelsFile.exists()) {
                println("GhidraDecompile: applying static data labels from " + staticLabelsPath);
                List<String[]> staticLabels = parseLabels(staticLabelsFile);
                int labeled = 0;
                for (String[] row : staticLabels) {
                    long rva = parseHex(row[0]);
                    if (rva == 0) continue;
                    Address addr = resolveAddress(rva, imageBase);
                    if (addr == null) continue;
                    try {
                        currentProgram.getSymbolTable().createLabel(
                            addr, row[1].trim(), SourceType.USER_DEFINED);
                        labeled++;
                    } catch (Exception ignored) {}
                }
                println("GhidraDecompile: labeled " + labeled + " / " + staticLabels.size() + " static addresses");
            } else {
                println("GhidraDecompile: static labels file not found (skipping): " + staticLabelsPath);
            }
        }

        // Set up decompiler
        DecompileOptions options = new DecompileOptions();
        DecompInterface decompiler = new DecompInterface();
        decompiler.setOptions(options);
        decompiler.setSimplificationStyle("decompile");
        decompiler.openProgram(currentProgram);

        ConsoleTaskMonitor taskMonitor = new ConsoleTaskMonitor();
        int ok = 0, fail = 0;

        for (int i = 0; i < rows.size(); i++) {
            String[] row = rows.get(i);
            // row: [addrStr, outputFile, lengthStr, typeName, memberName]
            String addrStr   = row[0].trim();
            String outPath   = row[1].trim();
            String typeName  = row.length > 3 ? row[3].trim() : "";
            String memberName = row.length > 4 ? row[4].trim() : "";

            long addrInt = parseHex(addrStr);
            if (addrInt == 0) {
                writeFile(outPath, "// Skipped: address is 0\n// Type: " + typeName + "\n// Member: " + memberName + "\n");
                fail++;
                continue;
            }

            // Try imageBase + rva first, then raw address
            Address ghidraAddr = resolveAddress(addrInt, imageBase);
            if (ghidraAddr == null) {
                String msg = "// ERROR: could not resolve address " + addrStr + "\n// Type: " + typeName + "\n// Member: " + memberName + "\n";
                writeFile(outPath, msg);
                fail++;
                println("  [" + (i+1) + "/" + rows.size() + "] ADDR_ERR " + typeName + " :: " + memberName);
                continue;
            }

            // Get or create function
            Function func = getFunctionAt(ghidraAddr);
            if (func == null) {
                func = createFunction(ghidraAddr, null);
            }
            if (func == null) {
                String msg = "// ERROR: no function at " + addrStr + "\n// Type: " + typeName + "\n// Member: " + memberName + "\n";
                writeFile(outPath, msg);
                fail++;
                println("  [" + (i+1) + "/" + rows.size() + "] NO_FUNC " + typeName + " :: " + memberName);
                continue;
            }

            // Decompile
            DecompileResults result = decompiler.decompileFunction(func, 60, taskMonitor);
            if (result.decompileCompleted()) {
                String cCode = result.getDecompiledFunction().getC();
                String header = "// Type  : " + typeName + "\n" +
                                "// Member: " + memberName + "\n" +
                                "// RVA   : " + addrStr + "\n" +
                                "// ────────────────────────────────────────\n\n";
                writeFile(outPath, header + cCode);
                ok++;
                println("  [" + (i+1) + "/" + rows.size() + "] OK  " + typeName + " :: " + memberName);
            } else {
                String errMsg = result.getErrorMessage();
                if (errMsg == null) errMsg = "unknown error";
                String msg = "// Decompilation failed\n// Type: " + typeName + "\n// Member: " + memberName +
                             "\n// RVA: " + addrStr + "\n// Error: " + errMsg + "\n";
                writeFile(outPath, msg);
                fail++;
                println("  [" + (i+1) + "/" + rows.size() + "] FAIL " + typeName + " :: " + memberName + " - " + errMsg);
            }
        }

        decompiler.dispose();
        println("\nGhidraDecompile: done. OK=" + ok + " FAIL=" + fail);
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private Address resolveAddress(long rva, long imageBase) {
        // Try absolute address (imageBase + rva)
        try {
            Address addr = currentProgram.getAddressFactory().getDefaultAddressSpace().getAddress(imageBase + rva);
            if (currentProgram.getMemory().contains(addr)) return addr;
        } catch (Exception ignored) {}

        // Fall back to raw value (file offset or already-absolute address)
        try {
            Address addr = currentProgram.getAddressFactory().getDefaultAddressSpace().getAddress(rva);
            if (currentProgram.getMemory().contains(addr)) return addr;
        } catch (Exception ignored) {}

        return null;
    }

    private static long parseHex(String s) {
        s = s.trim().replace("\"", "");
        if (s.isEmpty() || s.equals("0")) return 0;
        if (s.startsWith("0x") || s.startsWith("0X")) s = s.substring(2);
        try { return Long.parseUnsignedLong(s, 16); } catch (NumberFormatException e) { return 0; }
    }

    /**
     * Parses the manifest CSV, handling quoted OutputFile paths that may contain commas.
     * Returns list of String[] {addrStr, outputFile, lengthStr, typeName, memberName}.
     */
    private static List<String[]> parseManifest(File file) throws IOException {
        List<String[]> rows = new ArrayList<>();
        List<String> lines = Files.readAllLines(file.toPath());
        for (int i = 1; i < lines.size(); i++) {  // skip header
            String line = lines.get(i).trim();
            if (line.isEmpty()) continue;

            // Find quoted output path: Address,"path",Length,TypeName,MemberName
            int firstComma = line.indexOf(',');
            if (firstComma < 0) continue;
            String addrStr = line.substring(0, firstComma);
            String rest    = line.substring(firstComma + 1);

            String outPath, afterPath;
            if (rest.startsWith("\"")) {
                // Quoted path — find closing quote (handle "" escape inside)
                int end = 1;
                while (end < rest.length()) {
                    if (rest.charAt(end) == '"') {
                        if (end + 1 < rest.length() && rest.charAt(end + 1) == '"') {
                            end += 2; // escaped quote
                        } else {
                            break;
                        }
                    } else {
                        end++;
                    }
                }
                outPath   = rest.substring(1, end).replace("\"\"", "\"");
                afterPath = end + 1 < rest.length() ? rest.substring(end + 1) : "";
            } else {
                int nextComma = rest.indexOf(',');
                outPath   = nextComma >= 0 ? rest.substring(0, nextComma) : rest;
                afterPath = nextComma >= 0 ? rest.substring(nextComma) : "";
            }

            // afterPath starts with comma: ,Length,TypeName,MemberName
            String[] tail = afterPath.startsWith(",") ? afterPath.substring(1).split(",", 3) : new String[0];
            String lengthStr  = tail.length > 0 ? tail[0] : "0";
            String typeName   = tail.length > 1 ? tail[1] : "";
            String memberName = tail.length > 2 ? tail[2] : "";

            rows.add(new String[]{addrStr, outPath, lengthStr, typeName, memberName});
        }
        return rows;
    }

    private static void writeFile(String path, String content) throws IOException {
        File f = new File(path);
        File parent = f.getParentFile();
        if (parent != null && !parent.exists()) parent.mkdirs();
        try (FileWriter w = new FileWriter(f)) { w.write(content); }
    }

    /** Parses the labels CSV (header: RVA,Label). Returns list of {rva, label}. */
    private static List<String[]> parseLabels(File file) throws IOException {
        List<String[]> rows = new ArrayList<>();
        List<String> lines = Files.readAllLines(file.toPath());
        for (int i = 1; i < lines.size(); i++) { // skip header
            String line = lines.get(i).trim();
            if (line.isEmpty()) continue;
            int comma = line.indexOf(',');
            if (comma < 0) continue;
            rows.add(new String[]{line.substring(0, comma).trim(), line.substring(comma + 1).trim()});
        }
        return rows;
    }
}
